### PR TITLE
[oneDNN] Quantized MatMul with new API and features

### PR DESCRIPTION
**Pull request by mdfaijul**
_Reopened the closed PR https://github.com/tensorflow/tensorflow/pull/56792_

This PR introduces new API for quantized MatMul fusions. All fusions are implemented as single op _QuantizedMatMul with fused_ops attribute. This PR contains a number of features:

Weights can be quantized per-tensor or per-channel
MatMul + BiasAdd + Activation, where activation function can be any of {Relu, Relu6, LeakyRelu, Elu, Tanh, Sigmoid, GeluApproximate, GeluExact}
MatMul + BiasAdd + Add + Dequantize (when Add has same shape as the output of MatMul)
All fusions can have Dequantize as additional fusion.
All fusions can have Requantize as additional fusion, except the MatMul + BiasAdd + Add fusion.
When Dequantize is fused output can be float32
B matrix can be transposed.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h [hunk 1]
```diff
@@ -49,6 +49,38 @@ class MklTestingUtil {
...
     *tensor_max = max();
   }
+
+  // This utility function mimics Quantization of float/bfloat16 tensor with
+  // oneDNN backend QuantizeV2 operation. Since the op signature requires min
+  // and max values to be in float type, min_tensor and max_tensor should have
+  // their dtype set to DT_FLOAT.
+  template <typename T>
+  static Status GetQuantizationTensors(const Tensor& input, Tensor* output,
+                                     DataType out_type, const string mode,
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_kernel_util.h:59]: nit: indentation formatting

File: tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc [hunk 2]
```diff
@@ -268,16 +281,22 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, void, T> {
...
       std::shared_ptr<stream> cpu_stream;
-
       cpu_stream.reset(CreateStream(&eigen_tp, matmul_prim->GetEngine()));
 
       UserScratchPad<unsigned char> scratch_pad;
       scratch_pad.AllocateSPTensor(matmul_prim, ctx);
 
+      // Temporary tensor for scaled bias when op is quantized version.
+      Tensor temp_scaled_bias_tensor;
+      if (std::is_same<T2, qint8>::value) {
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc:295]: Should any other qint types be considered?

File: tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc [hunk 3]
```diff
@@ -322,34 +342,630 @@ class MklFusedMatMulOp : public MklDnnMatMulOpBase<T, void, T> {
...
+  int input_max_idx_ = 4;
+  int weight_min_idx_ = 5;
+  int weight_max_idx_ = 6;
+
+  struct PostOpInfo {
+    PostOpKind post_op_kind;
+    struct OperandInfo {
+      int idx = -1;  // Operand tensor index if needed by a post-op.
+      // Indices of min and max value tensors, if the operand is quantized.
+      gtl::InlinedVector<int, 4> min_max_indices;
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc:412]: These should all be replaced by `absl::InlinedVector` now.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 4]
```diff
@@ -102,8 +102,10 @@ struct MklDnnMatMulFwdParams {
   struct PostOpParam {
     string name;
     std::vector<float> param;
+    string partial_key = string("");
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:105]: Is there a reason this is explicitly initialized?

File: tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+                         const FusedOpsAndDims& fused_ops_and_dims,
+                         const GraphRunner& run_default,
+                         const FusedGraphRunner& run_fused, bool transpose_x,
+                         bool transpose_y, const double atol = 1e-5,
+                         // The following arguments are used by quantized fusion
+                         string input_quant_mode = "SCALED",
+                         string output_quant_mode = "SCALED",
+                         bool is_bias_quantized = false,
+                         bool is_perchannel = false, bool requantize = false) {
+    srand(1234);
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc:494]: Please don't rely on this.  Our random number generation sometimes changes, so this won't guarantee test stability, and likely leads to incorrect tolerances.

File: tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
+      float output_min = 1.0;
+      float output_max = 1.0 + std::numeric_limits<float>::epsilon();
+      if (requantize) {
+        T min;
+        T max;
+        MklTestingUtil::ComputeMinMax<T>(default_result, &min, &max);
+        output_min = static_cast<float>(min);
+        output_max = static_cast<float>(max);
+      }
+      // Run quantized fusion
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc:535]: nit: punctuation in/after comments, here and throughout.

**Review by cantonios — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc:295]: Among quantized types, kernel is registered only for qint8 type on T2.

File: tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc (hunk 5 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc]: Removed `srand`. Thanks!

File: tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc (hunk 6 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_fused_matmul_ops_test.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 4 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Removed explicit initialization.

File: tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc (hunk 3 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_kernel_util.h]: Fixed.

**Review by cantonios — APPROVED** [review]

**Pull request merged**